Add bowtie module

KevinMenden opened the pull request:
Added the bowtie module, consisting of `align` and `index`.
I've realised that `bowtie-build` actually doesn't need a GTF so could directly implement the tests without uploading one.

Both tests (`bowtie_index` and `bowtie_align`) are working.

Let me know what needs to be changed :)

Emiller88 reviewed (approved):
Tests look good to me and overall!

KevinMenden commented:
Well at some point we need a GTF anyway. For Bowtie though there's no workflow where it would need a GTF. 
Could still try to find one and upload it here - I can't properly test it then though.

Alternatively, I can make some tests for STAR which are still missing, and add a GTF to that PR then :)
As you like.

drpatelh commented on software/bowtie/align/main.nf:
```diff
@@ -0,0 +1,39 @@
+    path  index
+    
+    output:
+    tuple val(meta), path("*.sam")            , emit: sam
```
```suggestion
    tuple val(meta), path("*.sam"), emit: sam
```

drpatelh commented on software/bowtie/align/main.nf:
```diff
@@ -0,0 +1,39 @@
+    
+    output:
+    tuple val(meta), path("*.sam")            , emit: sam
+    tuple val(meta), path("*.out")            , emit: log
```
```suggestion
    tuple val(meta), path("*.out"), emit: log
```

drpatelh replied:
Mostly my aesthetic OCD but can we line these all up please?

drpatelh commented on software/bowtie/align/main.nf:
```diff
@@ -0,0 +1,39 @@
+    path  "bowtie.version.txt"                , emit: version
+
+    script:
+    def software   = getSoftwareName(task.process)
+    def prefix     = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
```
```suggestion
    def software  = getSoftwareName(task.process)
    def prefix    = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
    def unaligned = params.save_unaligned ? "--un ${prefix}.unmapped" : ''
```

drpatelh commented on software/bowtie/align/main.nf:
```diff
@@ -0,0 +1,39 @@
+    script:
+    def software   = getSoftwareName(task.process)
+    def prefix     = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    index_array = index.collect()
+    index = index_array[0].baseName - ~/.\d$/
```
```suggestion
```

drpatelh commented on software/bowtie/align/main.nf:
```diff
@@ -0,0 +1,39 @@
+    index_array = index.collect()
+    index = index_array[0].baseName - ~/.\d$/
+    """
+    bowtie $options.args  \\
+    --threads $task.cpus \\
+    ${index}  \\
+    -q ${reads} \\
+    --un ${prefix}.unAl > ${prefix}.sam 2> ${prefix}.out
+    bowtie --version | head -n 1 | cut -d" " -f3 > ${software}.version.txt
```
```suggestion
    INDEX=`find -L ./ -name "*.1.ebwt" | sed 's/.1.ebwt//'`
    bowtie \\
        --threads $task.cpus \\
        $options.args \\
        $INDEX \\
        -q ${reads} \\
	$unaligned \\
	> ${prefix}.sam 2> ${prefix}.out

    bowtie --version | head -n 1 | cut -d" " -f3 > ${software}.version.txt
```

drpatelh replied:
This is just me trying to free-style change this module so things will still need to be tested. Can we out BAM instead please? This will mean using a multi-tool container. See [HISAT2 module](https://github.com/nf-core/modules/blob/master/software/hisat2/align/main.nf#L56).

Also, will this work for paired-end data too?

KevinMenden replied:
Thanks - yeah no haven't tested for paired-end data yet, but will add tests for this as well!

drpatelh commented on software/bowtie/align/meta.yml:
```diff
@@ -0,0 +1,57 @@
+        description: |
+            Value for the Nextflow `publishDir` mode parameter.
+            Available: symlink, rellink, link, copy, copyNoFollow, move.
+    - enable_conda:
+        type: boolean
+        description: |
+            Run the module with Conda using the software specified
+            via the `conda` directive
```
Add `params.save_unaligned` here too.

drpatelh commented on software/bowtie/align/meta.yml:
```diff
@@ -0,0 +1,57 @@
+        description: Bowtie genome index files
+        pattern: "*.ebwt"
+output:
+    - index:
```
`index` isn't an `emit` output for this module.

drpatelh commented on software/bowtie/index/main.nf:
```diff
@@ -0,0 +1,31 @@
+    path fasta
+
+    output:
+    path "*.index*"          , emit: index
```
Spacing please ;)

drpatelh commented on software/bowtie/index/main.nf:
```diff
@@ -0,0 +1,31 @@
+    script:
+    def software  = getSoftwareName(task.process)
+    """
+    bowtie-build --threads $task.cpus ${fasta} ${fasta.baseName}.index
```
Can we output this as a directory instead. See [HISAT2 build](https://github.com/nf-core/modules/blob/79a0ce79ab9ab5d857f9a153d4c36d7dfc2a8d48/software/hisat2/build/main.nf#L51-L61).

drpatelh reviewed (changes requested):
Thanks @KevinMenden. Lots of niggly comments that need to be sorted for consistency, flexibility and standardisation.

KevinMenden commented:
Alright so most things are addressed. will also add BAM output

KevinMenden commented:
Alright made the samtools + bowtie=1.3.0 conttainer and added BAM output
also fixed a bug in the paired-end alignment that wasn't catched by the test

drpatelh commented:
Nice! Thanks. Can you add the Singularity image paths too like were added in https://github.com/nf-core/modules/pull/109. Will also need to update `meta.yml`. Just ping me on Slack when you are ready for a review.

KevinMenden commented:
Ah sorry forgot that these guys are back in the game now :) thanks for checking!

drpatelh commented on software/bowtie/align/main.nf:
```diff
@@ -0,0 +1,50 @@
+        2> ${prefix}.out \\
+        | samtools view $options.args2 -@ $task.cpus -bS -o ${prefix}.bam -
+
+    bowtie --version | head -n 1 | cut -d" " -f3 > ${software}.version.txt
```
```suggestion
    echo \$(bowtie --version 2>&1) | sed 's/^.*bowtie-align-s version //; s/ .*\$//' > ${software}.version.txt
```

drpatelh replied:
Will need to test it but more robust than a `cut`

drpatelh commented on software/bowtie/index/main.nf:
```diff
@@ -0,0 +1,35 @@
+    """
+    mkdir bowtie
+    bowtie-build --threads $task.cpus ${fasta} bowtie/${fasta.baseName}.index
+    bowtie --version | head -n 1 | cut -d" " -f3 > ${software}.version.txt
```
```suggestion
    echo \$(bowtie --version 2>&1) | sed 's/^.*bowtie-align-s version //; s/ .*\$//' > ${software}.version.txt
```

drpatelh commented on software/bowtie/align/meta.yml:
```diff
@@ -0,0 +1,65 @@
+        type: file
+        description: Output SAM file containing read alignments
+        pattern: "*.{sam}"
+    - version:
```
Add extra element here for `fastq` coming from `--un`.

drpatelh reviewed (changes requested):
Looking great @KevinMenden ! Few more things to address and we are good to go.

KevinMenden commented:
Any idea why the tests are suddenly failing? It can't find any of the output files - they are all prddouced though when trying locally.
Md5s also didn't change.

@Emiller88 maybe you have an idea?

KevinMenden commented:
Okay the suggestion of removing the `.index` from the bowtie index process introduced the error.
This was the missing from the output files of course, oversaw this in the `test.yml`

Alright so now only the EditorConfig doesn't like it

drpatelh reviewed (approved):
Beautiful 😍 Great work @KevinMenden and thank you!
